### PR TITLE
HTTP/3: Request and response data rates

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/dotnet/aspnetcore/issues/29702

### CODE REVIEW

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs [hunk 1]
```diff
@@ -192,6 +196,376 @@ public async Task ControlStream_RequestHeadersTimeoutMaxValue_ExpirationIsMaxVal
...
+            AdvanceClock(TimeSpan.FromSeconds(1));
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(TimeoutReason.WriteDataRate), Times.Once);
+
+            _mockTimeoutHandler.VerifyNoOtherCalls();
+        }
+        */
+
+        [Fact]
+        [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/21520")]
```
Comment (JamesNK) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs:278]: Copied from HTTP/2. I have fixed its flakey issue here - https://github.com/dotnet/aspnetcore/pull/32139

I will update this PR with that fix once it is merged.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs [hunk 2]
```diff
@@ -246,10 +251,15 @@ public void OnTimeout(TimeoutReason reason)
...
                     break;
-                case TimeoutReason.RequestHeaders: // Request header timeout is handled in starting stream queue
-                case TimeoutReason.KeepAlive:  // Keep-alive is handled by msquic
                 case TimeoutReason.ReadDataRate:
+                    Abort(new ConnectionAbortedException(CoreStrings.BadRequest_RequestBodyTimeout), Http3ErrorCode.InternalError);
+                    break;
                 case TimeoutReason.WriteDataRate:
+                    Log.ResponseMinimumDataRateNotSatisfied(_context.ConnectionId, traceIdentifier: null);
+                    Abort(new ConnectionAbortedException(CoreStrings.ConnectionTimedBecauseResponseMininumDataRateNotSatisfied), Http3ErrorCode.InternalError);
+                    break;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:260]: This seems about the same as HttpConnection.OnTimeout(). Do you think it's worth changing Http3ConnectionMiddleware to call through HttpConnection like HttpConnectionMiddleware does?

Then we could make Http3Connection an IRequestProcessor like Http1Connection and Http2Connection giving us greater consistency. HttpConnection itself is pretty small, so I think it be a fairly small change.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs [hunk 3]
```diff
@@ -192,6 +196,376 @@ public async Task ControlStream_RequestHeadersTimeoutMaxValue_ExpirationIsMaxVal
...
+                ignoreNonGoAwayFrames: false,
+                expectedLastStreamId: 8,
+                Http3ErrorCode.InternalError,
+                null);
+
+            _mockTimeoutHandler.VerifyNoOtherCalls();
+        }
+
+        /*
+         * Additional work around closing connections is required before response drain can be supported.
```
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs:242]: Is there an issue tracking this?

Comment (JamesNK) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs:232]: Added link to comment

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs [hunk 4]
```diff
@@ -133,6 +133,11 @@ public void StopProcessingNextRequest(bool serverInitiated)
 
                 if (Interlocked.CompareExchange(ref _gracefulCloseInitiator, initiator, GracefulCloseInitiator.None) == GracefulCloseInitiator.None)
                 {
+                    // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-5.2-11
+                    // An endpoint that completes a graceful shutdown SHOULD use the H3_NO_ERROR error code
+                    // when closing the connection.
+                    _errorCodeFeature.Error = (long)Http3ErrorCode.NoError;
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:139]: Should we just make NoError a long?

**Review by jkotalik — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs (hunk 4 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:139]: What is your reasonsing here? Is it about the cast?

Http3ErrorCode is an enum. I believe you always need to cast when going from an enum to a number.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:263]: Discussed offline. Will attempt in a follow-up PR.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs (hunk 4 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:139]: Yeah nvrm it is already a long.

**Pull request merged**